compiler/ballerina-lang/src/main/java/io/ballerina/compiler/api/impl/BallerinaSemanticModel.java

```diff
@@ -107,23 +103,14 @@ public BallerinaSemanticModel(BLangPackage bLangPackage, CompilerContext context
     @Override
     public Optional<Symbol> symbol(String srcFile, LinePosition position) {
         BLangCompilationUnit compilationUnit = getCompilationUnit(srcFile);
-        NodeResolver nodeResolver = new NodeResolver();
-        BLangNode node = nodeResolver.lookup(compilationUnit, position);
-
-        if (node instanceof IdentifiableNode) {
-            BSymbol symbol = (BSymbol) ((IdentifiableNode) node).getSymbol();
-            return Optional.ofNullable(SymbolFactory.getBCompiledSymbol(symbol, symbol.name.value));
-        } else if (node != null && (node.getKind() == USER_DEFINED_TYPE
-                || node.getKind() == NodeKind.UNION_TYPE_NODE
-                || node.getKind() == NodeKind.INTERSECTION_TYPE_NODE
-                || node.getKind() == NodeKind.VALUE_TYPE
-                || node.getKind() == NodeKind.BUILT_IN_REF_TYPE
-                || node.getKind() == NodeKind.CONSTRAINED_TYPE)) {
-            return Optional.ofNullable(
-                    SymbolFactory.createTypeDefinition(node.type.tsymbol, node.type.tsymbol.name.value));
+        SymbolFinder symbolFinder = new SymbolFinder();
+        BSymbol symbolAtCursor = symbolFinder.lookup(compilationUnit, position);
```

Comment:
Can we return an `Optional<BSymbol>` omitting the null check. Later in L113; we can use  a ternary for the return of `SymbolFactory.getBCompiledSymbol(...)` or `Optional.empty()`. As a further improvement we can return optional `Symbol` from the `SymbolFactory.getBCompiledSymbol(...)` as well.

Reply:
Not different from doing a null check right? And `SymbolFactory.getBCompiledSymbol()` should always be returning a `Symbol`, provided a valid `BSymbol`. Will update it to validate the `BSymbol`. Any filtering should be done before calling the factory method.